### PR TITLE
Improve test reliability

### PR DESCRIPTION
**Pull request by compulim**
### Fixed

-  Improved test reliability and adding snapshot to console of test harness in development mode, by [@compulim](https://github.com/compulim), in PR [#3891](https://github.com/microsoft/BotFramework-WebChat/pull/3891)

## Description

Now, we can show the image snapshot in console:

Console showing snapshot

## Specific Changes

- Group multiple `sendTab` and `sendShiftTab` into a single call
- `pageObjects.scrollToTop`
   - Before scrolling, wait for scroll to stabilize
      - If we scroll while the scroll is animating, our scroll may not land in precise position
   - After scrolling, wait for scroll to stabilize at the right scroll position
- After rendering Web Chat with a static transcript, wait for all activities to show up
- Add unit tests for `marshal.js` and `unmarshal.js`
- Fix an issue in RPC that it is not marshalling when resolve/reject
- Test harness running in development mode will take snapshot and draw it in the console log
- Taking a snapshot will wait until the page is stabilized:
   - Wait until all images are completed (either loaded or errored)
   - Wait until 2 consecutive screenshots look exactly the same
- Added a new option on `sendMessageViaSendBox`, `waitForNumResponse`, to wait for X number of responses from bot
   - `pageObjects.sendMessageViaSendBox('Hello, World!', { waitForNumResponse: 1 })` will wait for 1 response from bot
- Bump to [`react-scroll-to-bottom@4.1.1`](https://npmjs.com/package/react-scroll-to-bottom) for a race condition fix
   - https://github.com/compulim/react-scroll-to-bottom/pull/85

-  [x] I have added tests and executed them locally
-  [x] I have updated `CHANGELOG.md`
-  [x] ~I have updated documentation~

## Review Checklist

> This section is for contributors to review your work.

-  [x] ~Accessibility reviewed (tab order, content readability, alt text, color contrast)~
-  [x] ~Browser and platform compatibilities reviewed~
-  [x] ~CSS styles reviewed (minimal rules, no `z-index`)~
-  [x] ~Documents reviewed (docs, samples, live demo)~
-  [x] ~Internationalization reviewed (strings, unit formatting)~
-  [x] ~`package.json` and `package-lock.json` reviewed~
-  [x] ~Security reviewed (no data URIs, check for nonce leak)~
-  [x] Tests reviewed (coverage, legitimacy)

### CODE REVIEW

**Comment by corinagum** [comment]
FYI, I made some edits to your PR overview. LMK what you think :)

image

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 -  Fixes [#3812](https://github.com/microsoft/BotFramework-WebChat/issues/3812). Update `adaptiveCardHostConfig` to accessible text color-contrasts, by [@corinagum](https://github.com/corinagum) in PR [#3853](https://github.com/microsoft/BotFramework-WebChat/pull/3853)
 -  Fixes [#3816](https://github.com/microsoft/BotFramework-WebChat/issues/3816). De-bumped Node.js engine requirement for Direct Line Speech SDK to `>= 10.14.2` from `>= 12.0.0`, by [@compulim](https://github.com/compulim) in PR [#3854](https://github.com/microsoft/BotFramework-WebChat/pull/3854)
 -  Fixes [#3842](https://github.com/microsoft/BotFramework-WebChat/issues/3842). Updated [`husky`](https://npmjs.com/package/husky), [`lint-staged`](https://npmjs.com/package/lint-staged) and corresponding `precommit` scripts, by [@compulim](https://github.com/compulim), in PR [#3871](https://github.com/microsoft/BotFramework-WebChat/pull/3871)
+-  Improved test reliability and adding snapshot to console of test harness in development mode, by [@compulim](https://github.com/compulim), in PR [#3891](https://github.com/microsoft/BotFramework-WebChat/pull/3891)
```
Comment (corinagum) [inline CHANGELOG.md:37]:
```suggestion
-  Improved test reliability and added snapshots to console of test harness in development mode, by [@compulim](https://github.com/compulim), in PR [#3891](https://github.com/microsoft/BotFramework-WebChat/pull/3891)
```

File: packages/component/package.json [hunk 2]
```diff
@@ -69,7 +69,7 @@
     "react-film": "3.0.0",
     "react-redux": "7.2.3",
     "react-say": "2.0.2-master.ee7cd76",
-    "react-scroll-to-bottom": "4.1.0",
+    "react-scroll-to-bottom": "4.1.1-master.300a16c",
```
Comment (corinagum) [inline packages/component/package.json:72]: What's the plan for updating to an official release? 4.14? Could you file a bug and add it to the milestone?

Comment (corinagum) [reply packages/component/package.json:72]: I strongly prefer that the changes to `react-scroll-to-bottom` are also reviewed. Sync?

File: packages/test/harness/src/browser/globals/imageAsLog.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+async function getBlobURL(base64) {
```
Comment (corinagum) [inline packages/test/harness/src/browser/globals/imageAsLog.js:1]: exciting!

File: packages/test/harness/src/host/common/allImagesCompleted.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+const until = require('./until');
```
Comment (corinagum) [inline packages/test/harness/src/host/common/allImagesCompleted.js:1]: Just confirming, this runs automatically on every test, right?

File: packages/test/harness/src/host/common/host/sendTab.js [hunk 5]
```diff
@@ -1,7 +1,17 @@
...
-    return webDriver.actions().sendKeys(Key.TAB).perform();
+  return function sendTab(count = 1) {
+    if (!(count >= 1 && count <= 100)) {
+      throw new Error('First argument passed to sendTab() must between 1 and 100.');
+    }
+
+    let actions = webDriver.actions();
+
+    while (count--) {
+      actions = actions.sendKeys(Key.TAB);
```
Comment (corinagum) [inline packages/test/harness/src/host/common/host/sendTab.js:12]: This doesn't need an exit param?

File: packages/test/harness/src/host/dev/hostOverrides/error.js [hunk 6]
```diff
@@ -1,12 +1,11 @@
...
 
 const dumpLogs = require('../../common/dumpLogs');
-const override = require('../utils/override');
 const stripANSI = require('strip-ansi');
 
 // Send the error back to the browser console.
-module.exports = (webDriver, error) =>
-  override(error, async function error(error) {
+module.exports = webDriver =>
+  async function error(error) {
```
Comment (corinagum) [inline packages/test/harness/src/host/dev/hostOverrides/error.js:8]: Are there rules for Promise vs async/await?

**Review by corinagum — APPROVED** [review]
Approved, pending a few comments (and questions, haha)

File: packages/test/harness/src/host/common/allImagesCompleted.js (hunk 4 above)
Comment (compulim) [reply packages/test/harness/src/host/common/allImagesCompleted.js:1]: This run before every snapshot is taken. 🎉

File: packages/test/harness/src/host/common/host/sendTab.js (hunk 5 above)
Comment (compulim) [reply packages/test/harness/src/host/common/host/sendTab.js:12]: Exit parameter?

- `count` must be 1 to 100
- If `count` is `0` *before* decrement, then stop

File: packages/test/harness/src/host/dev/hostOverrides/error.js (hunk 6 above)
Comment (compulim) [reply packages/test/harness/src/host/dev/hostOverrides/error.js:8]: I guess you mean, `then`/`catch` vs. `async`/`await`.

99% of the time, use `async`/`await`.

1%, IMO, is the main entrypoint:

```js
async function main() {
  // This is main entrypoint
}

// You can't use await at root level
main().catch(err => {
  console.error(err);
  process.exit(-1);
});
```

File: packages/component/package.json (hunk 2 above)
Comment (compulim) [reply packages/component/package.json]: It's production now, 4.1.1. It's mainly this PR https://github.com/compulim/react-scroll-to-bottom/pull/85.

I added the test harness plus some "happy path" tests, thus the diff is pretty long, [`v4.1.0...v4.1.1`](https://github.com/compulim/react-scroll-to-bottom/compare/v4.1.0...v4.1.1).

**Comment by compulim** [comment]
Saw this, thanks!

image

**Pull request merged**